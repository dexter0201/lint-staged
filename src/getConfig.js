/* eslint no-console: 0 */
const { intersection, merge } = require('lodash')

const defaultConfig = {
  concurrent: true,
  gitDir: '.',
  globOptions: {
    matchBase: true,
    dot: true
  },
  linters: {},
  renderer: 'update',
  verbose: false
}

/**
 * For a given configuration object that we retrive from .lintstagedrc or pacakge.json
 * construct a full configuration with all options set.
 *
 * This is a bit tricky since we support 2 different syntxes: simple and full
 * For simple config, only the `linters` configuration is provided.
 *
 * @param {Object} config
 * @returns {{
 *   concurrent: boolean, gitDir: string, linters: Object, renderer: string, verbose: boolean
 * }}
 */

module.exports = function getConfig(config) {
  // Check if the config is "simple" i.e. doesn't contains any of full config keys
  if (
    !('linters' in config) &&
    intersection(Object.keys(defaultConfig), Object.keys(config)).length === 0
  ) {
    return merge(defaultConfig, {
      linters: config
    })
  }
  // Output config in verbose mode
  // if (verbose) console.log(completeConfig)
  return merge(defaultConfig, config, {
    renderer: config.verbose ? 'verbose' : 'update'
  })
}
