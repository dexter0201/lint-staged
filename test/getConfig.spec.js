/* eslint no-console: 0 */

import stripColors from 'strip-ansi'
import getConfig from '../src/getConfig'

describe('getConfig', () => {
  it('should return config with defaults', () => {
    expect(getConfig({})).toMatchSnapshot()
  })

  it('should set verbose', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        verbose: false
      })
    )

    expect(
      getConfig({
        verbose: false
      })
    ).toEqual(
      expect.objectContaining({
        verbose: false
      })
    )

    expect(
      getConfig({
        verbose: true
      })
    ).toEqual(
      expect.objectContaining({
        verbose: true
      })
    )
  })

  it('should set concurrent', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        concurrent: true
      })
    )

    expect(
      getConfig({
        concurrent: false
      })
    ).toEqual(
      expect.objectContaining({
        concurrent: false
      })
    )

    expect(
      getConfig({
        concurrent: true
      })
    ).toEqual(
      expect.objectContaining({
        concurrent: true
      })
    )
  })

  it('should set renderer based on verbose key', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        renderer: 'update'
      })
    )

    expect(
      getConfig({
        verbose: false
      })
    ).toEqual(
      expect.objectContaining({
        renderer: 'update'
      })
    )

    expect(
      getConfig({
        verbose: true
      })
    ).toEqual(
      expect.objectContaining({
        renderer: 'verbose'
      })
    )
  })

  it('should set gitDir', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        gitDir: '.'
      })
    )

    expect(
      getConfig({
        gitDir: '..'
      })
    ).toEqual(
      expect.objectContaining({
        gitDir: '..'
      })
    )
  })

  it('should set linters', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        linters: {}
      })
    )

    expect(
      getConfig({
        '*.js': 'eslint'
      })
    ).toEqual(
      expect.objectContaining({
        linters: {
          '*.js': 'eslint'
        }
      })
    )

    expect(
      getConfig({
        linters: {
          '*.js': ['eslint --fix', 'git add'],
          '.*rc': 'jsonlint'
        }
      })
    ).toMatchSnapshot()
  })

  it('should deeply merge configs', () => {
    expect(
      getConfig({
        globOptions: {
          nocase: true
        }
      })
    ).toEqual(
      expect.objectContaining({
        globOptions: {
          nocase: true,
          matchBase: true,
          dot: true
        }
      })
    )
  })

  xit('should output config to stdout in verbose mode', () => {
    const stdout = []
    console.log = jest.fn().mockImplementation(input => {
      stdout.push(JSON.stringify(stripColors(input)))
    })
    getConfig({ verbose: true })
    expect(console.log).toHaveBeenCalledTimes(1)
    expect(stdout).toMatchSnapshot()
  })
})
