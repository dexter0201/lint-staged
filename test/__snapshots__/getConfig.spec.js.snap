// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig should no 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not  1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not m 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mu 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not muta 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate  1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate o 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate or 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate ori 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate orig 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate origi 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate origin 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate origina 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate original  1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate original 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate original confi 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should not mutate original config 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should return config with defaults 1`] = `
Object {
  "concurrent": true,
  "gitDir": ".",
  "globOptions": Object {},
  "linters": Object {},
  "renderer": "update",
  "verbose": false,
}
`;

exports[`getConfig should set linters 1`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;

exports[`getConfig should set linters 2`] = `
Object {
  "concurrent": true,
  "gitDir": "..",
  "globOptions": Object {},
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "verbose": true,
}
`;
